src/init.cpp

```diff
@@ -793,6 +793,31 @@ void InitParameterInteraction(ArgsManager& args)
             LogInfo("parameter interaction: -onlynet excludes IPv4 and IPv6 -> setting -dnsseed=0\n");
         }
     }
+
+    // We drop UPnP support but kept the arg as hidden for now to display a friendlier error to user who have the
+    // option in their config, and upgrade the setting to -natpmp. TODO: remove for version 30.0.
+    if (args.IsArgSet("-upnp")) {
+        bool overridden = args.SoftSetBoolArg("-natpmp", true);
+        LogWarning("Option '-upnp' is set but UPnP support was dropped in version 29.0.%s",
+                overridden ? " Setting '-natpmp' instead." : "");
+    }
+
+    // If dynamic settings are enabled, also replace it in settings.json.
+    fs::path path;
+    if (args.GetSettingsPath(&path, /* temp= */ false)) {
+        common::SettingsValue value = args.GetPersistentSetting("upnp");
+        if (!value.isNull()) {
+            LogWarning("Option '-upnp' has a value in settings.json, replacing it with `-natpmp`");
```

Comment:
`in bitcoin.conf or settings.json, replacing it with -natpmp in settings.json` 

When it's in `bitcoin.conf` the warning will get repeated every startup, which seems fine.

Reply:
Yup. That's intentional. The first warning can happen as many times as needed, the json manipulation should ideally only happen once.